twoliter/embedded/imghelper

```diff
@@ -108,6 +110,214 @@ mkfs_data_xfs() {
+  pesigcheck -i "${shim}" -n 0 -c "${SBKEYS}/db.cer"
+
+  # Remove the original key if necessary, then sign the MOK manager as well.
+  if [[ "$(pesign -i "${mokm}" -l)" == "No signatures found." ]]; then
+    mv "${mokm}" "${mokm}.unsigned"
+  else
+    pesign -i "${mokm}" -o "${mokm}.unsigned" -u 0 -r
+  fi
+  pesign -i "${mokm}.unsigned" -o "${mokm}.signed" -s "${code_sign_key[@]}"
+  mv "${mokm}.signed" "${mokm}"
+  rm "${mokm}.unsigned"
+  pesign -i "${mokm}" -l
+  pesigcheck -i "${mokm}" -n 0 -c "${SBKEYS}/vendor.cer"
+
+  # Remove the original vendor key if necessary, replace the embedded gpg
+  # public key, then sign grub with the vendor key.
+  if [[ "$(pesign -i "${grub}" -l)" == "No signatures found." ]]; then
+    mv "${grub}" "${grub}.unsigned"
+  else
+    pesign -i "${grub}" -o "${grub}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${grub}.unsigned" \
+    --file-alignment 4096 \
+    --update-section ".pubkey=${SBKEYS}/config-sign.pubkey"
+  pesign -i "${grub}.unsigned" -o "${grub}.signed" -s "${code_sign_key[@]}"
+  mv "${grub}.signed" "${grub}"
+  rm "${grub}.unsigned"
+  pesign -i "${grub}" -l
+  pesigcheck -i "${grub}" -n 0 -c "${SBKEYS}/vendor.cer"
+}
+
+uefi_dummysign() {
+  local arch shims shim grubs grub
+  arch="${1:?}"
+  shims=(boot*.efi)
+  shim="${shims[0]}"
+  grubs=(grub*.efi)
+  grub="${grubs[0]}"
+
+  # Generate a zero-sized certificate in the expected format.
+  cert_table /dev/null "${SBKEYS}/vendor.obj"
+
+  # Replace the embedded vendor certificate with the zero-sized one, which
+  # shim will ignore when Secure Boot is disabled.
+  "${arch}-objdumpcopy" "${shim}" \
+    --update-section ".vendor_cert=${SBKEYS}/vendor.obj"
+
+  # Remove the embedded gpg public key to disable GRUB's signature checks.
+  "${arch}-objdumpcopy" "${grub}" \
+    --file-alignment 4096 \
+    --remove-section ".pubkey"
+}
+
+uefi_copycert() {
+  local efi_image
+  efi_image="${1:?}"
+  mcopy -i "${efi_image}" "${SBKEYS}"/db.{crt,cer} ::/EFI/BOOT
+}
+
+vmlinuz_sbsetup() {
+  local vmlinuz
+  vmlinuz="${1:?}"
+  if [[ "$(pesign -i "${vmlinuz}" -l)" == "No signatures found." ]]; then
+    mv "${vmlinuz}" "${vmlinuz}.unsigned"
+  else
+    pesign -i "${vmlinuz}" -o "${vmlinuz}.unsigned" -u 0 -r
+  fi
+  pesign -i "${vmlinuz}.unsigned" -o "${vmlinuz}.signed" -s "${code_sign_key[@]}"
+  mv "${vmlinuz}.signed" "${vmlinuz}"
+  rm "${vmlinuz}.unsigned"
+  pesign -i "${vmlinuz}" -l
+  pesigcheck -i "${vmlinuz}" -n 0 -c "${SBKEYS}/vendor.cer"
+}
+
+vmlinuz_hmac() {
+  local vmlinuz
+  vmlinuz="${1:?}"
+  openssl sha512 -hmac FIPS-FTW-RHT2009 -hex "${vmlinuz}" |
+    awk -v vmlinuz="${vmlinuz}" '{ print $2 "  " vmlinuz }' \
+      >"${vmlinuz%/*}.${vmlinuz##*/}.hmac"
+}
+
+sign_grub() {
+  local grub_cfg grub_cfg_sig
+  grub_cfg="${1:?}"
+  grub_cfg_sig="${grub_cfg}.sig"
+  [[ -f "${grub_cfg_sig}" ]] && rm "${grub_cfg_sig}"
+  gpg --batch --no-tty --detach-sign "${grub_cfg}"
+  gpg --batch --no-tty --verify "${grub_cfg_sig}" "${grub_cfg}"
```

Comment:
The reason for the change in syntax here is when `--batch` is not used `gpg` can make assumptions about the matching data files from a single argument. This is discouraged and instead we should always specify both explicitly. Since we're not interfacing with `gpg` interactively anyway, I just went ahead and set all the `gpg` commands to `--batch --no-tty`, which results in more useful errors when something goes wrong.